src/bio/smd/smd_pool.c

```diff
@@ -147,55 +160,65 @@ smd_pool_del_tgt(uuid_t pool_id, uint32_t tgt_id)
 }
 
 static struct smd_pool_info *
-smd_pool_alloc_info(struct d_uuid *id, struct smd_pool *pool)
+smd_pool_alloc_info(struct d_uuid *id, struct smd_pool *pools)
 {
 	struct smd_pool_info	*info;
+	enum smd_dev_type	 st;
 	int			 i;
 
 	D_ALLOC_PTR(info);
 	if (info == NULL)
 		return NULL;
 
-	D_ALLOC_ARRAY(info->spi_tgts, SMD_MAX_TGT_CNT);
-	if (info->spi_tgts == NULL) {
-		smd_pool_free_info(info);
-		return NULL;
-	}
-	D_ALLOC_ARRAY(info->spi_blobs, SMD_MAX_TGT_CNT);
-	if (info->spi_blobs == NULL) {
-		smd_pool_free_info(info);
-		return NULL;
+	for (st = SMD_DEV_TYPE_DATA; st < SMD_DEV_TYPE_MAX; st++) {
+		D_ALLOC_ARRAY(info->spi_tgts[st], SMD_MAX_TGT_CNT);
+		if (info->spi_tgts[st] == NULL) {
+			smd_pool_free_info(info);
+			return NULL;
+		}
+		D_ALLOC_ARRAY(info->spi_blobs[st], SMD_MAX_TGT_CNT);
+		if (info->spi_blobs[st] == NULL) {
+			smd_pool_free_info(info);
+			return NULL;
+		}
+		info->spi_blob_sz[st] = pools[st].sp_blob_sz;
+		info->spi_tgt_cnt[st] = pools[st].sp_tgt_cnt;
+
+		for (i = 0; i < pools[st].sp_tgt_cnt; i++) {
+			info->spi_tgts[st][i] = pools[st].sp_tgts[i];
+			info->spi_blobs[st][i] = pools[st].sp_blobs[i];
+		}
 	}
 
 	D_INIT_LIST_HEAD(&info->spi_link);
 	uuid_copy(info->spi_id, id->uuid);
-	info->spi_blob_sz = pool->sp_blob_sz;
-	info->spi_tgt_cnt = pool->sp_tgt_cnt;
-	for (i = 0; i < info->spi_tgt_cnt; i++) {
-		info->spi_tgts[i] = pool->sp_tgts[i];
-		info->spi_blobs[i] = pool->sp_blobs[i];
-	}
 	return info;
 }
 
 int
 smd_pool_get_info(uuid_t pool_id, struct smd_pool_info **pool_info)
 {
 	struct smd_pool_info	*info;
-	struct smd_pool		 pool;
+	struct smd_pool          pools[SMD_DEV_TYPE_MAX];
+	enum smd_dev_type	 st;
 	struct d_uuid		 id;
 	int			 rc;
 
+	memset(pools, 0, sizeof(struct smd_pool) * SMD_DEV_TYPE_MAX);
 	uuid_copy(id.uuid, pool_id);
 	smd_db_lock();
-	rc = smd_db_fetch(TABLE_POOL, &id, sizeof(id), &pool, sizeof(pool));
-	if (rc) {
-		D_ERROR("Fetch pool "DF_UUID" failed: "DF_RC"\n",
-			DP_UUID(&id.uuid), DP_RC(rc));
-		goto out;
+	for (st = SMD_DEV_TYPE_DATA; st < SMD_DEV_TYPE_MAX; st++) {
+		rc = smd_db_fetch(TABLE_POOLS[st], &id, sizeof(id), &pools[st], sizeof(pools[st]));
+		/* META and WAL are optional */
+		rc = (rc == -DER_NONEXIST && st > SMD_DEV_TYPE_DATA) ? 0 : rc;
```

Comment:
then we should break the loop right? I know it's not supposed to happen, but what will happen if we have MD but no WAL, should we report error?

Reply:
Breaking the loop here means failing smd_pool_get_info() in configurations with DATA blobs/targets only. MD with no WAL is indeed an error config but it could be dicovered by checking the smd_pool_info struct returned.